### DISCUSSION TITLE
How to aggregate Errors efficiently like you can aggregate Validation

### DISCUSSION THREAD

**Discussion by kgkoutis** [louthy/language-ext#1070] [Q&A] [answered]
First of all congrats for the work here it's literally fantastic. I haven't found an answer to my issue so here it goes.

I would like to somehow "keep tabs" of the Errors produced in some kind of sequence without the computation hard stopping like it does with a `@catch` . A `Validation` has the possibility to do that since the `Fail` factory method can accept a `Seq<FAIL>` so that afterwards you can fold over it if you want and can write expressions like so: 

(example from wiki)
```csharp
var numberV = DigitsOnly(number) | MaxStrLength(16)(number);
```

Following another example from wiki we see the following piece of code.
```csharp	
static Eff<RT, Unit> askUser =>
		repeat(from ln in Console<RT>.readLine
			   from _1 in guard(notEmpty(ln), UserExited)
			   from _2 in guardnot(ln == "sys", () => throw new SystemException())
			   from _3 in guardnot(ln == "err", () => throw new Exception())
			   from _4 in Console<RT>.writeLine(ln)
			   select unit)
			 | @catch(UserExited, unit);
```

How is it possible to keep repeating the computation if for example the line is empty - and thus ask again for more input - but without handling it immediately with a `@catch` function? Ideally I would like to see and **handle all exceptions later**, say in the wiki's example when I am about the outer `Main(String[] args)` method? Out there I will be able to examine every exception/error that got gathered in the fail sequence. Is this possible? 

PS: Of course maybe the `Error` need to be renamed from `UserExited` to `UserError`, since the computation will continue to go on..

**Comment by louthy** [comment, marked as answer]
Monadic operations force an order, and must exit as soon as there's an error.  If you think about it it's obvious:

```c#
    from x in broken
    from y in working
    select x + y;
```
Clearly we can't have a situation where `broken` fails and then leaves `x` in an uninitialised state.  

The way to achieve capture of multiple errors is to use applicatives, which are independent terms in an expression.  Those independent terms can all be run and all errors collected, but it's a bit more awkward to do with C#.  

I have just [added support for a multi-error `Error` type, and applicative functor support to `Aff` and `Eff`](https://github.com/louthy/language-ext/releases/tag/v4.3.0-beta).  

It's a bit late here now, and I'll admit to having not fully read your request properly yet.  However It might be worth having a read of the release notes to see if it helps you.

**Comment by kgkoutis** [comment]
@louthy After a lot of pondering and soul searching what I am really going after, I believe you brought all the pieces to play by adding applicative behavior of `Aff` and `Eff` with support on `ManyErrors`. I believe the type I am looking for is `Aff<RT, (ManyErrors, A)>` so that the final computation is really fault tolerant with `.Apply` that will fold over independent computations keeping also all errors, `.Bind` that allows binding previous computations to "string" the results into a composed monad and `@catch` that can allow restorative actions by something like 
```csharp

from _1 in askUser 
                | @catch(ex => ex is SystemException, err => { return SuccessAff<RT, (ManyErrors, int)>(combineTuples((err, 0), _prevComp); })
from _2 in Console<RT>.writeLine("goodbye")
```
where `0` is a restorative result (for ints), `_prevComp` is also a `(ManyErrors, A)` tuple as the the result of the previous computation and `combineTuples` knows how to add them together. I hope that `Error.Many(Error.Many("error one"), Error.Many("error two"));` is not some form of `Error.Many<Error.Many>` and needs to be flattened.

I am marking it as answer.
